### ISSUE TITLE
GLM-5.3-Flash: which Transformers version produced the published HF logit parity?

### ISSUE THREAD

**Issue by waynel96** [NVIDIA-NeMo/Automodel#3940]
The [GLM-5.3-Flash model-coverage page](https://docs.nvidia.com/nemo/automodel/model-coverage/vision-language-models/thudm/glm-5-3-flash) reports a Hugging Face logit parity result and, in the same page, states which Transformers version the training path needs. I could not reproduce the parity from the environment the page prescribes, because the Transformers-native implementation of this model does not exist at the pinned version.

Filing this as a question rather than a bug — the most likely explanation is that the parity was produced with a newer or pre-release Transformers than the training path pins, in which case the page may just need a note.

### What the page says

[Supported Training Path, L44-L45](https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/docs/model-coverage/vlm/thudm/glm-5-3-flash.mdx#L44-L45):

> The implementation runs on the repository's pinned `transformers==5.12.1`; users do not need Transformers 5.16, a nightly Transformers build, or `trust_remote_code`.

[Numerical Validation → Hugging Face Logit Parity, L122-L133](https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/docs/model-coverage/vlm/thudm/glm-5-3-flash.mdx#L122-L133):

> A four-layer model containing KDA, KDA, KDA, and DSA was compared with the Transformers reference at sequence length 4,096 over the full 154,880-token vocabulary. All 135 expected tensors loaded with no missing, unexpected, or shape-mismatched keys.

| Metric | Result |
|---|---:|
| Mean KL, HF to NeMo AutoModel | 0.0002477953 |
| P95 KL | 0.0005324535 |
| Max KL | 0.0271428488 |
| Logits cosine similarity | 0.9999251899 |
| Top-1 token agreement | 98.33984375% |

### What I measured

`transformers.models.glm5_next` is absent from 5.12.1, 5.15.1 and 5.16.0, and first appears in [**5.17.0**](https://github.com/huggingface/transformers/tree/v5.17.0/src/transformers/models/glm5_next). Checked by listing `transformers/models/*` inside each released wheel:

- **5.15.1** — `glm`, `glm4`, `glm46v`, `glm4_moe`, `glm4_moe_lite`, `glm4v`, `glm4v_moe`, `glm_image`, `glm_moe_dsa`, `glm_ocr`, `glmasr`, `glmga`, `xglm` — no `glm5_next`
- **5.16.0** — same set, no `glm5_next`
- **5.17.0** — adds `configuration_glm5_next.py`, `modeling_glm5_next.py`, `modular_glm5_next.py`, `processing_glm5_next.py`, and the image/video processors

The [`zai-org/GLM-5.3-Flash` repository](https://huggingface.co/zai-org/GLM-5.3-Flash/tree/main) contains no `.py` files, so the `trust_remote_code` path is not available either.

On [`nvcr.io/nvidia/nemo-automodel`](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo-automodel) `@sha256:c21c0c3b82f68fcccfded4bf9fe11b2d53df1c9a969e9e883a1562ffca6294f6`, `transformers.__version__` is `5.12.1` and `glm5_next` is likewise absent.

### Why this matters (and what is unaffected)

**Unaffected:** running and training the model. AutoModel ships [its own `glm5_next` implementation](https://github.com/NVIDIA-NeMo/Automodel/tree/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/glm5_next), so it does not need Transformers to carry the model. I loaded a four-layer language-only slice successfully under the image above with a `9a463234` source overlay — 110 expected keys, zero missing, zero unexpected, FP8 dequantization and grouped-expert aggregation both clean.

**Affected:** the reference. A fixed-input logit comparison needs a second independent implementation, and in the prescribed environment there is not one available, so a reader following the page cannot reproduce the published parity numbers.

For contrast, the sibling [Qwen3.8-Flash-Next page](https://docs.nvidia.com/nemo/automodel/model-coverage/large-language-models/qwen/qwen3-8-flash-next) compares against an SGLang reference and makes no Transformers version claim, so this looks specific to the GLM page rather than systemic.

### Follow-up: the two implementations are not structurally identical

After filing, I built the reference side in a separate virtualenv (`transformers==5.17.0`, same torch and container) and constructed the same four-layer slice on both sides. They differ in two places:

| | AutoModel `glm5_next` | Transformers 5.17.0 `glm5_next` |
|---|---|---|
| Four-layer expected keys | 457 (**110** language + 347 vision) | 451 (**104** language + 347 vision) |
| Grouped experts | `experts.gate_and_up_projs` / `experts.down_projs` | `experts.gate_up_proj` / `experts.down_proj` |
| mHC | `attn_hc.fn`, `attn_hc._fp32_params.base`, `attn_hc._fp32_params.scale` | `attn_hc.fn`, `attn_hc.base`, `attn_hc.scale` |
| KDA short conv | `self_attn.q_conv1d`, `k_conv1d`, `v_conv1d` (three) | `self_attn.conv1d` (one, fused) |
| KDA decay | `self_attn._fp32_params.A_log` | `self_attn.forget_gate.A_log` |

The 110 vs 104 difference is exactly the short-conv decomposition: three KDA layers times two extra conv tensors each. The two should be numerically equivalent, but this is the kind of structural divergence that bears on question 3 below.

Also worth noting for anyone attempting a reproduction: only **63** of the 451 Transformers-side keys match the released checkpoint's names exactly, so the Transformers side performs its own renaming and expert grouping on load as well. A reproduction therefore has to go through `from_pretrained` on both sides rather than a manual `load_state_dict`, or the comparison stops being between two independent implementations.

### Questions

1. Which Transformers version (or pre-release build) produced the published HF logit parity?
2. Given the training path pins 5.15.1, would it make sense for the page to point readers at a separate reference environment for reproducing the parity?
3. Does "All 135 expected tensors loaded" count AutoModel-side expected keys? A four-layer language-only stub built from the released config gives **110** on the AutoModel side and **104** on the Transformers side, so I may be constructing a different stub than the one measured. If the intended stub differs, I would be glad to know how.

### Environment

```
AutoModel source   9a4632347323fbfee9ba1ae1a323669ef74fc5fb (overlay on the image below)
image              nvcr.io/nvidia/nemo-automodel@sha256:c21c0c3b82f68fcccfded4bf9fe11b2d53df1c9a969e9e883a1562ffca6294f6
torch              2.13.0a0+8145d630e8.nv26.06, CUDA 13.3
GPU                NVIDIA B200, compute capability 10.0
model revision     zai-org/GLM-5.3-Flash @ eb9eb208eb0d988989d07a6a12d0fdeb5f52574a
backend            attn=sdpa, experts=torch, dispatcher=torch, rms_norm=torch_fp32
reference venv     transformers==5.17.0, same torch and container
```

Related: #3939 fixes two smaller factual issues on the same page.

**Comment by HuiyingLi** [comment]
Hi @waynel96 Thank you for using Automodel and reproducing the logits! Also thank you for updating the docs of this model!
Could you please try: https://github.com/huggingface/transformers/commit/dabae5fcb924a8eece0e727b627ca5f050b40d40 
w.r.t number of keys, 135 count is the Automodel side expected keys. The number of keys from AM vs HF does not need to match, we have state dict adapters that might alter param layout.
For the reproduction: first 4 text layers (KDA KDA KDA DSA) and vision_config.depth=1, should map to 110 language tensors+25 vision tnsor.

**Comment by waynel96** [comment]
Thanks @HuiyingLi — that resolves the tensor-count question completely.

I had been building the stub with the full vision tower, so I was seeing **110 language keys + 347 vision keys**. With `vision_config.depth=1` giving 25 vision tensors, 110 + 25 = 135 lines up exactly, and my 110 language count matches yours. Good to have confirmation that AM-vs-HF key counts are not expected to match — I had already seen the adapter fuse the 288 per-expert tensors into `experts.gate_and_up_projs` / `experts.down_projs`, so that is consistent.

On the linked commit: I may be misreading it, but `dabae5fcb924a8eece0e727b627ca5f050b40d40` looks like *"Create the continuous batching CPU group with local synchronization (#48302)"* from 2026-08-26, touching only `generation/continuous_batching/distributed.py`, and it is already an ancestor of `v5.17.0`. Did you mean a different commit?

The reason I ask is that my original question is narrower than the tensor counts: **which Transformers provides the `glm5_next` reference implementation the parity was measured against.** Listing `transformers/models/*` inside each released wheel, `glm5_next` is absent from 5.12.1, 5.15.1 and 5.16.0, and first appears in **5.17.0** — while the page states the training path runs on the repository's pinned `transformers==5.12.1` and that users do not need 5.16. Both statements are individually fine (AutoModel ships its own `glm5_next`, so training genuinely does not need Transformers to carry the model), but a reader following the page cannot obtain the reference side needed to reproduce the published KL/cosine/top-1 numbers.

If the parity was produced with a pre-release or git build of Transformers, a one-line note to that effect on the page would be enough — happy to send that as a follow-up PR if you tell me the version.

**Comment by waynel96** [comment]
**transformers 5.17.0's `glm5_next` cannot load the released `zai-org/GLM-5.3-Flash` checkpoint — the parameter names differ, including one structural difference.**

Following up on the reproduction recipe above: I installed `transformers==5.17.0` and built the stub exactly as described — first 4 text layers (`linear_attention` ×3 + `deepseek_sparse_attention`, matching what `layer_types` gives) and `vision_config.depth=1`. Construction works: `AutoConfig` resolves to `Glm5NextConfig` (`model_type=glm5_next`) and `AutoModelForImageTextToText.from_config` gives `Glm5NextForConditionalGeneration`, 9.715 B parameters, **129 state-dict keys = 103 language + 25 vision**. The vision count matches your 25 exactly.

The problem is the names. For `model.language_model.layers.0` the checkpoint has 29 tensors and the 5.17.0 stub expects 24, and **only 13 names match**:

| 5.17.0 expects | released checkpoint has |
|---|---|
| `attn_hc.{base,fn,scale}` | `hc_attn_{base,fn,scale}` |
| `ffn_hc.{base,fn,scale}` | `hc_ffn_{base,fn,scale}` |
| `self_attn.forget_gate.{A_log,dt_bias,f_a_proj,f_b_proj}` | `self_attn.{A_log,dt_bias,f_a_proj,f_b_proj}` |
| `self_attn.conv1d.weight` (one tensor) | `self_attn.{q,k,v}_conv1d.weight` (three) |

The last row is not a rename — it needs the three projections concatenated, and the expected order is not documented.

There is also no built-in remap to fall back on: `Glm5NextForConditionalGeneration._checkpoint_conversion_mapping` is absent, and across the whole `transformers/models/glm5_next/` package the strings `q_conv1d` and `hc_attn` appear zero times. So `from_pretrained` on the released checkpoint yields missing and unexpected keys rather than a loaded model.

This is what I am actually trying to get at with the original question. The published HF logit parity (mean KL 0.00025, max 0.027, top-1 98.34%) cannot have been produced by vanilla 5.17.0 against the released checkpoint, so either a different Transformers revision whose naming matches the checkpoint was used, or there is a conversion step I am missing. Could you point at whichever it is? If a converter exists, that mapping — particularly the conv1d concatenation order — is the missing piece; if it is a version, the model-coverage page saying so would save the next person this detour.

Happy to send a docs PR once I know which.

**Comment by HuiyingLi** [comment]
Hi @waynel96 ,
Thank you for clarifying. 
The HF commit: yeah I started the work after this model is already merged into transformers.  So the commit is just a random commit that already has the reference impl, even tho the commit itself is irrelevant to this model. If there is any change on either side AM(esp if new kernel is added) or HF that these numbers drift a lot, I am happy to investigate/look at your fix :)

**Comment by HuiyingLi** [comment]
Hi, w.r.t the ckpt loading in HF, could you please take a look at https://github.com/huggingface/transformers/blob/856157a2f3e9594954310df18fdccc31ffddebe9/src/transformers/conversion_mapping.py#L552-L597
